Added saving mechanism to cleaning txt files.

mewlware opened the pull request:
Every time you finished a cleaning stage, you can always go back to any previous stages that has been cleaned, depending on what number you input (the terminal will help you by showing what number of stages you can choose to go back to!). However if you go back to a previous stage, you'd have to redo subsequent stages regardless of whether or not you've cleaned it before to keep the cleaning mechanism John made.

Example: John has finished with stage 0, 1, 2, 3, but decided something is wrong with the cleaning in stage 1. John can input 1 to restart the stage, but the results from stage 2 and 3 is wiped out clean. After restarting stage 1, John has to restart stage 2 and 3 as well.

TODO:
1. Add written-to-disk saving mechanism if program crashes
2. Add undo mechanism for individual cleaning steps per stage
3. Maybe fix: change law_stages header to enums

mewlware committed:
Added saving mechanism to cleaning txt files.

Every time you finished a cleaning stage, you can always go back to any previous stages that has been cleaned, depending on what number you input (the terminal will help you show what number of stages you can choose to go back to!). However if you go back to a previous stage, to keep the cleaning mechanism John made you'd have to redo subsequent stages regardless of whether or not you've cleaned it before.

Example: John has finished with stage 0, 1, 2, 3, but decided something is wrong with the cleaning in stage 1. John can input 1 to restart the stage, but the results from stage 2 and 3 is wiped out clean. After restarting stage 1, John has to restart stage 2 and 3 as well.

TODO:
1. Add written-to-disk saving mechanism if program crashes
2. Add undo mechanism for individual cleaning steps per stage
3. Maybe fix: change law_stages header to enums

mewlware committed:
Fix bugs and better user-friendliness

Fixed bug where cleaning goes on indefinitely.

Fixed bug whereas if you are finished with the last cleaning stage, you can't go back to cleaning stage and press n.

Better user support on terminal for ending the cleaning stages.

mewlware commented:
Fixed bug where cleaning goes on indefinitely.

Fixed bug whereas if you are finished with the last cleaning stage, you can't go back to cleaning stage and press n (the only option was to restart the last cleaning stage regardless of whether you think it's necessary).

Better user support on terminal for ending the cleaning stages.

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
-    return law
+    return law_stages[-1][1]
 
+#def delete_subsequent_law_stages(n_stage: int, law_stages: List[list[list]]) -> List[list[list]]:
+#    for i in range(n_stage, len(law_stages)):
+#        law_stages[i][1].clear()
+#    return law_stages
```
Do you wanna leave this out of this PR?

mewlware replied:
Sure!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+
+        if cleaning_progress['started'] == True:
+
+            print("---------------")
```
can use `print_line` here <- I made a util just for this

there's also `print_dashed_line`

mewlware replied:
Yeah, noticed this after I requested the PR 😄 (because I was using the parser code from last week before pulling from origin today). Will change it today!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+            cleaning_progress['started'] = True
+            print_section_header("STARTING CLEANING")
+
+        #while int(user_input_stage) not in range(len(law_stages)) or user_input_stage != 'n':
```
delete?

mewlware replied:
Yep 👍

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+            print(f"Pick a stage {pick_number} or")
+            user_input_stage = input(f"enter {pick_n} to go to the next stage: ")
+        
+        elif cleaning_progress['started'] == False:
+            cleaning_progress['started'] = True
+            print_section_header("STARTING CLEANING")
```
Can we move this outside of the while loop and just putting this at the top before the while loop. That way you can get rid of the `cleaning_progress['started']` variable entirely (then `cleaning_progress['finished']` can just be `has_finished_cleaning` or you can even get rid of that too and just check if the current stage is larger than `len(law_stages)`

mewlware replied:
Yep! 👍

mewlware replied:
Oh wait! I think I added this so that when you first start cleaning, it immediately starts the first stage w/o asking you for input. I agree that it's probably not the best way to do it though.

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
-
-    e.g a real line ending in '2 / 43' in UU 18 2017
+    Preps for cleaning stages for potential save in a dict of stages
+    TODO(Mel):  1. Maybe change law_stages headers into enums?
```
yes pls :) you can see `parser_types.py` for how to make an enum

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+    
+    while cleaning_progress['finished'] == False:
+        
+        user_input_stage = '0'
```
nitpick: would it be easier if you just made this an int? that way you don't have to keep doing `int(user_input_stage) in multiple places`? dunno if it'd actually be better, just a suggestion

mewlware replied:
Ah it's simply because I think it's more user friendly to allow user to input 'n' instead of a number to go to the next stage w/o confusing it with a stage number. I do agree it's not DRY if I have to keep doing int(user_input_stage) everywhere though. My (admittedly quick) solution was just to save the int-casted input into a variable after the logic finished checking for non numeric values. But I'm open to other solutions!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+        '''Checks whether stage n - 1 has content (and thus can proceed to cleaning stage n)'''
+        if user_input_stage == 'n':
+            cleaned_stages = len(law_stages)
+            for i, stage in enumerate(law_stages):
+                if len(stage[1]) == 0:
+                    user_input_stage = i
+                    break
+                else:
+                    cleaned_stages -= 1
 
-    save_law_to_file(law, 'b.txt')
+            if cleaned_stages == 0:
+                user_input_stage = -1
```
Wouldn't this be easier & simpler if you just had a variable that keeps track of what stage the user is currently on? :) Something like:

```
if user_input == 'n':
      # insert logic for handling last stage
      current_stage = current_stage + 1
```

right now we have to loop through all of `law_stages` every time just to get the next stage (not that the runtime complexity matters, BUT I do think it makes the logic unnecessarily complicated)

mewlware replied:
Aha, I tried that. In my head this weirdly seemed more complex 😖. Will do so however!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
-    law = clean_maybe_list_items(law)
-
-    save_law_to_file(law, 'c.txt')
+        elif user_input_stage.isnumeric() == False:
```
nitpick: btw you can also do `elif not user_input_stage.isnumeric():` 

both patterns are ok (1 is not better than the other); just w/ to point out in case you didn't know

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
-    TODO(johnamadeo): Fix "Pasal 38 B" REMOVE THE SPACE WITH REGEX
-    '''
-    law = clean_split_pasal_number(law)
+        else:
+            print(f"{colored('Invalid input.', 'red')}")
```
This is just for general error handling and we never expect ourselves to actually get here right? 

If so, `raise Exception('Some message')` instead of just printing to console.

Crashing early & often when we know we can't handle the error is the best way to eliminate silent failures

mewlware replied:
I see. Okay!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+            '''
+            Deal with list indexes. See clean_maybe_list_item for more.
+            '''
+            law_stages[1][1] = clean_maybe_list_items(law_stages[0][1])
```
suggestion:

do 
`law_stages[int_user_input_stage][1] = clean_maybe_list_items(law_stages[int_user_input_stage-1][1])`

instead

if we add new stages before/after and cleaning maybe list items becomes the n-th stage where is no longer 1, you'll have to remember to update all the array indexes manually. (Or if someone else does, they might not know to update it or miss it).

Same applies for everything else w/ the same pattern below.

mewlware replied:
Yep :). Missed this last night, thanks!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
+            '''
+            law_stages[5][1] = insert_perubahan_quotes(law_stages[4][1])
+
+        elif int_user_input_stage == -1:
```
I don't really like using -1 as a convention for done here. It's more stylistic, but I think it's much, much clearer if it were just an enum (e.g `CleaningStage.DONE`) esp. if the person reading has no prior exposure to the code

also, this might just be me, but when I see something set to -1 conventionally I expect this to be some kind of error case

mewlware replied:
Makes sense!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +318,156 @@ def clean_law(law: List[str]) -> List[str]:
     '''
-    law = list(filterfalse(ignore_line, law))
-    law = clean_squashed_page_numbers(law)
+    law_stages = [['clean_squashed_page_numbers',[]], ['clean_maybe_list_items',[]],
```
Can we change this data structure to something like
```
law_stages = [
      { 
             'stage': CleaningStage.CLEAN_SQUASHED_PAGE_NUMBERS,
             'law': [],
      },
     ...
]
```

In general, pick the data structure that makes the logic easiest to read, especially for 
- a) other people, and 
- b) you 5 months from now when you've forgotten how this PR was written 😛 

W/ the current data structure, the code is littered with stuff like `law_stages[1][1]` where I have no idea what the second array index means at all. If we had something like `law_stages[1]['law']` then that would at least give more of an idea. 

For example, now when I read `len(law_stages[int_user_input_stage - 1][1]) == 0` I know that the len = 0 check means you're checking if the law has any lines i.e we already cleaned that stage.

I think the TLDR here is dictionaries are your friend haha (I'm writing a slightly longer comment here since I noticed you did this too in the first parser PR you made)

mewlware replied:
Yeah I agree! I actually started this using dict for the same reason you mentioned above, but I changed it into a list for _some_ reason I can't remember now... Probably because I forgot OrderedDict existed (I wanted to preserve the ordering of the stages, but if I use enums as stage headings, would it matter?).

mewlware replied:
Ignore my previous comment (for multiple reasons). Will implement as you've stated 👍

JohnAmadeo reviewed (changes requested):
1) Logic looks good to me!
2) I pulled the branch & tried it out - UI looks great. I could understand what was going on + can already see this saving a lot of time.
3) I think there are opportunities to simplify the logic & make it clearer to read. Most of the comments revolve around this.

Once the comments are addressed, should be good to approve & merge! \(^-^)/ Nice job on your first parser PR!

mewlware committed:
Updates saving mechanism

Includes:

1. Enum implementation of cleaning stages
2. Dictionary implementation of cleaning stages : resulting law
3. Documentation on saving mechanism
4. Cleaning up saving mechanism logic based on John's requests

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +403,158 @@ def clean_law(law: List[str]) -> List[str]:
 
-    save_law_to_file(law, 'b.txt')
+    next_cleaning_stage = 1
+    len_CleaningStageOrder = len(CleaningStageOrder)
```
nit: don't break naming conventions

since all the variables are in underscore_case, this should be `len_cleaning_stage_order` instead of being a mashup of camelCase and underscore_case haha :)

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -48,6 +49,41 @@
+CLEANING_STAGES: Dict[CleaningStageOrder, Dict[str, List[str]]] = {
+    CleaningStageOrder.CLEAN_SQUASHED_PAGE_NUMBERS:
+        {
+            'law_cleaned': []
```
nit: can we change this to `cleaned_law`?

this is another one of those stylistic things, but I like my variable names to actually sound like English nouns (e.g you wouldn't say "sashimi is fish raw" you'd say "sashimi is raw fish" i.e raw_fish > fish_raw)

^ (btw, a lot of your other variable names already follow this implicitly e.g you added `next_cleaning_stage` not `cleaning_stage_next`, `new_law` not `law_new`

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +403,158 @@ def clean_law(law: List[str]) -> List[str]:
+            print(f"{colored('Invalid number. No stage with that number.', 'red')}")
+            continue
 
+        elif (int_input_stage != 1 and len(CLEANING_STAGES[CleaningStageOrder(int_input_stage - 1)]['law_cleaned']) == 0):
```
do you wanna break this up so the line is shorter? like maybe pull out `CleaningStageOrder(int_input_stage - 1)` into its own variable first?

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +403,158 @@ def clean_law(law: List[str]) -> List[str]:
+
+        elif int_input_stage in range(2, len_CleaningStageOrder + 1):
+            try:
+                new_law = clean_law_at_stage(int_input_stage, CLEANING_STAGES[CleaningStageOrder(int_input_stage - 1)]['law_cleaned'])
```
this line feels too long 😆 

it's ok to use multiple lines hahaha (can also consider pulling out CleaningStageOrder(int_input_stage - 1) into separate var first
```
new_law = clean_law_at_stage(
         int_input_stage, 
         CLEANING_STAGES[CleaningStageOrder(int_input_stage - 1)]['law_cleaned']
)
```

mewlware replied:
Instead of pulling out CleaningStageOrder(int_input_stage - 1) I took the liberty to change CLEANING_STAGE[CleaningStageOrder(int_pick_stage) into current_stage and CLEANING_STAGE[CleaningStageOrder(int_pick_stage - 1) into previous_stage. It seems to me like the code is still understandable, although with another layer of naming abstraction now haha. Do you agree?

*int_input_stage is also changed into int_pick_stage just to be consistent with other naming conventions.

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +403,158 @@ def clean_law(law: List[str]) -> List[str]:
+            print_line()
+            
+            pick_number = colored('number', 'blue')
+            pick_n = colored('n', 'blue')
```
Can we not use 'n' and pick another letter?

I noticed that when I tried it out, my brain had already hardwired 'n' = 'no' since that's how 'n' is used for every other prompt in our cleaning logic

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -318,56 +403,158 @@ def clean_law(law: List[str]) -> List[str]:
+            print("(n) to undo to specific cleaning stages.")
+            print_yes_no()
+
+            input_finish = ''
```
same nit as above: `finish_input` instead of `input_finish`

btw I noticed above this is not how you name your input variables
```
            pick_number = colored('number', 'blue')
            pick_n = colored('n', 'blue')
```

maybe to be consistent w/ the snippet above you can name this `finish_parsing`

anyway, TLDR is be consistent w/ variable names

JohnAmadeo reviewed (approved):
Looks good to me! Nice job Mel - the logic is definitely cleaner & more extensible/future-proof.

I pointed out some stylistic nits to fix, and after those are fixed should be good to merge! As the name suggests - "nitpicks" - they can feel relatively minor, but I've been fairly zealous about pointing them out since you're still relatively new to programming & I think over time fixing these will benefit you 🙂

mewlware commented:
Reasonable comments! It's fine, making code easier to mentally parse feels like a valuable skill :). Will commit changes today.

mewlware committed:
Change variable names & a bit of code structure

Implemented John's comments on variable names.
